core/src/avm2/value.rs

```diff
@@ -0,0 +1,287 @@
+        .get(index.0 as usize - 1)
+        .cloned()
+        .ok_or_else(|| format!("Unknown uint constant {}", index.0).into())
+}
+
+pub fn abc_double(file: &AbcFile, index: Index<f64>) -> Result<f64, Error> {
+    if index.0 == 0 {
+        return Ok(NAN);
+    }
+
+    file.constant_pool
+        .doubles
+        .get(index.0 as usize - 1)
+        .cloned()
+        .ok_or_else(|| format!("Unknown double constant {}", index.0).into())
+}
+
+/// Retrieve a string from an ABC constant pool, yielding `""` if the string
+/// is the zero string.
+pub fn abc_string(file: &AbcFile, index: Index<String>) -> Result<String, Error> {
+    if index.0 == 0 {
+        return Ok("".to_string());
+    }
+
+    file.constant_pool
+        .strings
+        .get(index.0 as usize - 1)
+        .cloned()
+        .ok_or_else(|| format!("Unknown string constant {}", index.0).into())
+}
+
+/// Retrieve a string from an ABC constant pool, yielding `None` if the string
+/// is the zero string.
+///
+/// This function still yields `Err` for out-of-bounds string constants, which
+/// should cause the runtime to halt. `None` indicates that the zero string is
+/// in use, which callers are free to interpret as necessary (although this
+/// usually means "any name").
+pub fn abc_string_option(file: &AbcFile, index: Index<String>) -> Result<Option<String>, Error> {
+    if index.0 == 0 {
+        return Ok(None);
+    }
+
+    file.constant_pool
+        .strings
+        .get(index.0 as usize - 1)
+        .cloned()
+        .map(Some)
+        .ok_or_else(|| format!("Unknown string constant {}", index.0).into())
+}
+
+/// Retrieve a default value as an AVM2 `Value`.
+pub fn abc_default_value<'gc>(
+    file: &AbcFile,
+    default: &AbcDefaultValue,
+) -> Result<Value<'gc>, Error> {
+    match default {
+        AbcDefaultValue::Int(i) => abc_int(file, *i).map(|v| v.into()),
+        AbcDefaultValue::Uint(u) => abc_uint(file, *u).map(|v| v.into()),
+        AbcDefaultValue::Double(d) => abc_double(file, *d).map(|v| v.into()),
+        AbcDefaultValue::String(s) => abc_string(file, s.clone()).map(|v| v.into()),
+        AbcDefaultValue::True => Ok(true.into()),
+        AbcDefaultValue::False => Ok(false.into()),
+        AbcDefaultValue::Null => Ok(Value::Null),
+        AbcDefaultValue::Undefined => Ok(Value::Undefined),
+        AbcDefaultValue::Namespace(ns)
+        | AbcDefaultValue::Package(ns)
+        | AbcDefaultValue::PackageInternal(ns)
+        | AbcDefaultValue::Protected(ns)
+        | AbcDefaultValue::Explicit(ns)
+        | AbcDefaultValue::StaticProtected(ns)
+        | AbcDefaultValue::Private(ns) => {
+            Namespace::from_abc_namespace(file, ns.clone()).map(|v| v.into())
+        }
+    }
+}
+
+impl<'gc> Value<'gc> {
+    pub fn as_object(&self) -> Result<Object<'gc>, Error> {
+        if let Value::Object(object) = self {
+            Ok(*object)
+        } else {
+            Err(format!("Expected Object, found {:?}", self).into())
```

Comment:
It seems to be only `null` and `undefined` should error out here, the rest should box. I guess that's coming in a later PR?

Reply:
That will be in the `avm2-primitives` branch, a child branch of this one. That currently has most of the coercions and abstract equality, but I still need to implement abstract comparison and value boxing.